src/libsyntax/parse/parser.rs

```diff
@@ -5401,23 +5401,26 @@ impl<'a> Parser<'a> {
             }
             let mut err = self.diagnostic().struct_span_err(id_sp,
                 "cannot declare a new module at this location");
-            let this_module = match self.directory.path.file_name() {
-                Some(file_name) => file_name.to_str().unwrap().to_owned(),
-                None => self.root_module_name.as_ref().unwrap().clone(),
-            };
-            err.span_note(id_sp,
-                          &format!("maybe move this module `{0}` to its own directory \
-                                     via `{0}/mod.rs`",
-                                    this_module));
+            if id_sp != syntax_pos::DUMMY_SP {
+                let path = self.sess.codemap().span_to_filename(id_sp);
+                let path = if path.ends_with(".rs") {
```

Comment:
I think it would be better to always remove the extension here, e.g. 
```rust
let path = Path::new(self.sess.codemap().span_to_filename(id_sp)).file_stem().unwrap();
```